### PR TITLE
[build] Split stdlib to it's own package.

### PR DESCRIPTION
**Pull request by ejgallego**
We introduce a new package structure for Coq:

- `coq-core`: Coq's OCaml tools code and plugins
- `coq-stdlib`: Coq's stdlib [.vo files]
- `coq`: meta-package that pulls `coq-{core,stdlib}`

This has several advantages, in particular it allows to install Coq
without the stdlib which is useful in several scenarios, it also open
the door towards a versioning of the stdlib at the package level.

The main user-visible change is that Coq's ML development files now
live in `$lib/coq-core`, for compatibility in the regular build we
install a symlink and support both setups for a while.

Note that plugin developers and even `coq_makefile` should actually
rely on `ocamlfind` to locate Coq's OCaml libs as to be more robust.

There is a transient state where we actually look for both
`$coqlib/plugins` and `$coqlib/../coq-core/plugins` as to support
the non-ocamlfind plus custom variables.

This will be much improved once #13617 is merged (which requires this
PR first), then, we will introduce a `coq.boot` library so finally
`coqdep`, `coqchk`, etc... can share the same path setup code.

IMHO the plan should work fine.

### CODE REVIEW

**Comment by rgrinberg** [comment]
> However this may break 3rd party applications, as now the library name will be coq-base.kernel etc

Not necessarily. In dune, we already have a `deprecated_library_name` stanza to facilitate moving libraries between packages painlessly.

In any case, this suggestion looks quite good.

File: coq.opam
Comment (LasseBlaauwbroek) [inline coq.opam:24]: This should probably not be here

**Comment by ejgallego** [comment]
Anyone willing to shepherd this, note that #13655 tries to do the same. So IMHO it is worth going ahead.

**Comment by Zimmi48** [comment]
I think a lot of people would be in favor of this split, but it is interesting to consider what will be the most backward-compatible path going forward. I prefer the virtual `coq` package, including an explicit `coq-stdlib` package over the solution proposed by #13655. However, AFAICS, this PR doesn't yet implement such a virtual `coq` package, does it?

Also, is there any chance to use @rgrinberg's proposal to give even more backward-compatibility?

> Not necessarily. In dune, we already have a `deprecated_library_name` stanza to facilitate moving libraries between packages painlessly.

**Comment by ejgallego** [comment]
> I think a lot of people would be in favor of this split, but it is interesting to consider what will be the most backward-compatible path going forward. I prefer the virtual `coq` package, including an explicit `coq-stdlib` package over the solution proposed by #13655. However, AFAICS, this PR doesn't yet implement such a virtual `coq` package, does it?

If it doesn't it is an oversight, it should do it. This PR is basically ready except for fixing the CI.

> Also, is there any chance to use @rgrinberg's proposal to give even more backward-compatibility?

Can try, but not critical, given than we are not stable for the ML API so people in 8.14 can just move `coq -> coq-core`.

**Comment by rgrinberg** [comment]
> Anyone willing to shepherd this, note that #13655 tries to do the same. So IMHO it is worth going ahead.

Would it make sense to get rid of the makefile based build system and fully transition to dune first? I'm guessing that it will be much more difficult to maintain backwards compat well with both build systems.

**Comment by ejgallego** [comment]
> Would it make sense to get rid of the makefile based build system and fully transition to dune first? I'm guessing that it will be much more difficult to maintain backwards compat well with both build systems.

There is #13617 , so indeed we will merge that one first hopefully.

**Comment by ejgallego** [comment]
Some minor nits pending, but overall this should be ready for review.

**Comment by ejgallego** [comment]
> Some minor nits pending, but overall this should be ready for review, in particular #13617 could make great use of this PR.

I'm happy with the state of this, modulo the TODO in the PR description but IMHO it should work.

Needs a shepherd.

File: clib/cPath.mli
Comment (gares) [inline clib/cPath.mli:18]:
```suggestion
```

File: clib/cPath.mli
Comment (gares) [inline clib/cPath.mli:12]: Can't you make it opaque right now? (you are introducing the type, nobody is relying on its definition)

File: coq-stdlib.opam
Comment (gares) [inline coq-stdlib.opam:11]: This description needs to be updated.

File: coq-core.opam
Comment (gares) [inline coq-core.opam:11]: This description needs to be updated.

File: test-suite/misc/printers.sh
Comment (gares) [inline test-suite/misc/printers.sh:4]:
```suggestion
```

File: test-suite/misc/printers.sh
Comment (gares) [inline test-suite/misc/printers.sh:8]:
```suggestion
```

**Review by gares — COMMENTED** [review]
If I install just `coq-core` does `coqtop` work? where is the prelude? Shouldn't coq-core depend on coq-stdlib?

**Comment by ejgallego** [comment]
> If I install just `coq-core` does `coqtop` work? where is the prelude? Shouldn't coq-core depend on coq-stdlib?

It works if you pass `-noinit`, for example coq-core + hott is fully functional.

File: clib/cPath.mli
Comment (ejgallego) [reply clib/cPath.mli:12]: I tried that some time ago, was far from trivial, and I was using dune `Path.t` API as a reference.

So indeed that is a good thing to do, whether to do in this PR or later I don't mind; but would introduce quite a few `Path.from_string`.

**Comment by ejgallego** [comment]
Thanks for pointing the leftovers / debug code, will fix after the CI run.

**Comment by gares** [comment]
> It works if you pass `-noinit`, for example coq-core + hott is fully functional.

Then I think it should be stated very clearly in the package description. I find it very weird that a package provides a binary that does not work unless you specify a very technical option. IMO you should make the error message from `coqlib()` mention that as well.

**Comment by ejgallego** [comment]
> Then I think it should be stated very clearly in the package description. I find it very weird that a package provides a binary that does not work unless you specify a very technical option. IMO you should make the error message from `coqlib()` mention that as well.
> 
> The alternative is to have a virtual package (in the debian sense) but I don't know if opam has it. I mean something like `coq-prelude` which is both provided by `coq-stdlib` but also `hott-stdlib` so that you get at least one of the two installed and hence your binary works.

Note that we already have the `coq` package, so I'd expect only expert users to install `coq-core` only, not sure how verbose about such low-level options we should be.

opam can do it's own virtual-style packages using or, but I am also not sure we would benefit from another package [whereas the split seems useful as it has been suggested independently by 4 or 5 people already]

Anyways I'm open to any kind of setup / documentation, so let's gather feedback and hopefully we have quite a bit of time to refine for 8.14.

**Comment by LasseBlaauwbroek** [comment]
This will be extremely useful for me. Thanks for your efforts!

File: coq-core.opam
Comment (silene) [inline coq-core.opam:27]: Why was a dependency on `num` added?

Also, should the dependency on `dune` be declared `build` only? I do not understand why users should be forced to recompile Coq (and all the depending Coq libraries) whenever one updates Dune. (I am pretty sure I will never update Dune if I know it will cause the recompilation of all the Coq libraries.)

Comment (Zimmi48) [reply coq-core.opam:27]: > Also, should the dependency on `dune` be declared `build` only? I do not understand why users should be forced to recompile Coq (and all the depending Coq libraries) whenever one updates Dune. (I am pretty sure I will never update Dune if I know it will cause the recompilation of all the Coq libraries.)

IIRC Dune installs `dune-project` files which contains the Dune version. Dune tries as hard as it can to be backward-compatible but if you downgrade your Dune version after having installed some Dune package, then you break your installation. Anyway, this led to removing the `{ build }` annotation from every `dune` dependency in the opam-repository.

File: Makefile.dune
Comment (Zimmi48) [inline Makefile.dune:59]:
```suggestion
	@echo "  - coq-core: base Coq package, toplevel compilers, tools, no stdlib, no GTK"
```

File: Makefile.install
Comment (Zimmi48) [inline Makefile.install:167]: I guess this is the compatibility line. Why is the `rm -f` needed?

File: coq-core.opam
Comment (Zimmi48) [reply coq-core.opam:11]: Indeed, this description should make it clear that this is an internal package not fit for direct installation by most users.

File: coq-core.opam
Comment (Zimmi48) [inline coq-core.opam:25]: In the Coq Call, it was said that this lower bound could be made even lower (for the pure OCaml coq-core package). But now that I think of it, this would require a separate `dune-project` file for `coq-core` and for `coq-stdlib`, since this is the file that specifies the minimal Dune version to use. That's a shame because I'm not even sure that this is a supported use case.

File: coq.opam
Comment (Zimmi48) [inline coq.opam:27]: Shouldn't you remove all build steps to make this a proper virtual package? And remove `dune` from the list of dependencies?

File: coq-core.opam
Comment (silene) [reply coq-core.opam:27]: > Dune installs `dune-project` files which contains the Dune version.

Did you mean `dune-package`? Also, as far as I can tell, these files do not contain the version of Dune (but they contain the language version, which seems sensible to me).

> if you downgrade your Dune version after having installed some Dune package, then you break your installation

I have a hard time understanding why "downgrading" is a scenario worth making the life of every user painful. So yes, I guess I will never upgrade Dune, if it forces me to recompile every single package.

**Review by Zimmi48 — COMMENTED** [review]
The Nix error is probably due to the test-suite being run *after* installation. This use case might be broken by the current state of the PR.

File: coq-core.opam
Comment (Zimmi48) [reply coq-core.opam:27]: Right. I don't know how frequently this happens. You should have a look at the relevant issues in the opam-repository for the context (I don't have the pointers, sorry).

Comment (ejgallego) [reply coq-core.opam:27]: `num` seems like a bad rebase, thanks for catching it [I didn't do myself a super careful review until I fix the CI first]

`dune` cannot be `build` these days indeed, actually there is a very long discussion about that, as both dune-packages files and build actually are problematic, people is aware of it and trying to find a solution [I guess for dune 3.0] I suggest to discuss in the proper channel.

Note that if you enable compilation cache, and you upgrade dune, it should be expected that the recompilation time for packages building with dune to be close to zero. Not that this is a solution, but indeed changes the point of view a bit when build rules are effectively memoized.

File: coq-core.opam
Comment (silene) [reply coq-core.opam:25]: Due to #13437, I would on the contrary say that the bound should be put even higher, i.e., 2.8. Because it certainly does not work for me (unless something changed since I reported that bug).

Comment (Zimmi48) [reply coq-core.opam:25]: This bug should be closed now.

Note that the issue with 2.8 is that it is very recent and not yet packaged in Debian / Ubuntu, which would break @JasonGross's ppa, which is currently important for several projects' CI.

The plan laid out by @ejgallego during the call yesterday was that Dune would be required only to build the OCaml parts of Coq and there would be a choice of a coq_makefile-based or a Dune-based build for the standard library. If in addition, we could require different Dune lower bounds for the two parts (but I'm worried of this not being technically feasible), this plan would present the nice additional property of not forcing any recent Dune version upon our users.

**Comment by SkySkimmer** [comment]
Would it make sense to have a package for the prelude between coq-core and coq-stdlib?

**Comment by silene** [comment]
I have been wondering as well but I concluded that it was not worth the effort. If some users have a need for the prelude, then they might also need some other files from the standard library, e.g., `Arith` or `List`. So, it feels like splitting just the prelude is not worth it. We might reconsider when we start splitting the standard library into larger chunks, e.g., `Reals`, `Fsets`.

**Comment by Alizter** [comment]
Like @silene, I feel as though splitting prelude is really about splitting up the stdlib. It's not a given that everybody wants to use the same coq prelude for starters, and perhaps making the stdlib more modular so people can pick and chose parts of it is a more useful thing to do.

File: coq-core.opam
Comment (ejgallego) [reply coq-core.opam:25]: Yes we can uncouple the versions but that would require using different `dune-project` files for coq-core and coq-stdlib, so that'd be a further refinement.

Comment (Zimmi48) [reply coq-core.opam:25]: I think this worth it, but indeed, it can be done later. @JasonGross: you have no problem with Dune 2.5 becoming a requirement to build Coq, right?

Comment (JasonGross) [reply coq-core.opam:25]: Ah, hold on, I should have looked into this more carefully earlier.  I've managed to package Dune 2.7 with OCaml 4.08.1 on all of the versions of Ubuntu that I care about, but I'm concerned that the 10x slowdown @ppedrot found at https://github.com/ocaml/ocaml/issues/7826#issuecomment-625342743 will hit fiat-crypto.  So it looks like I may need to move to 4.11.1 (can someone confirm that this version of OCaml does not suffer from this issue?), and I haven't yet packaged 4.11.1 for the various distributions I care about.  (And I may run into snags, as I did in packaging 4.08, such as OCaml being incompatible with the distributed version of binutils.)

Comment (Zimmi48) [reply coq-core.opam:25]: What about using Dune 2.7 compiled with OCaml 4.08.1 but staying on OCaml 4.07.1 for compiling Coq? Is this doable in your environment? If yes, that would be the safest solution.

Comment (ejgallego) [reply coq-core.opam:25]: Indeed, OCaml 4.07 was release almost 3 years ago so it is safe start preparing the move from that version, as the GC parameter bug that was creating some problem has been finally solved, it is likely the recommended version becomes 4.12.0 soon.

We already discussed it, but it seems to me that using Debian packages to track Coq's master is gonna always be quite a bit of work, and will get worse in the future as document managers and other tooling are allowed to depend in even more packages. IMHO other solutions such as Nix / Docker / Opam are much preferable and tend to work out of the box. It makes little sense for the Coq team to have to worry about the complexities of Debian packaging as it is really a distribution issue, not related at all to Coq development itself so hard to justify to spend significant cycles on.

Comment (JasonGross) [reply coq-core.opam:25]: [OCaml 4.11.1 does not even build on Groovy (Ubuntu 20.10), because many tests fail](https://launchpadlibrarian.net/525230922/buildlog_ubuntu-groovy-i386.ocaml_4.11.1-4.1~groovy~ppa1_BUILDING.txt.gz) with things like:
```
Running tests from 'tests/regression/pr3612' ...
 ... testing 'pr3612.ml' with 1 (native) => failed (The file /<<PKGBUILDDIR>>/testsuite/_ocamltest/tests/regression/pr3612/pr3612/ocamlopt.byte/ocamlopt.byte.output was expected to be empty because there is no reference file /<<PKGBUILDDIR>>/testsuite/tests/regression/pr3612/pr3612.compilers.reference but it is not:
========================================
/usr/bin/ld: pr3612.o: warning: relocation in read-only section `.text'
/usr/bin/ld: warning: creating DT_TEXTREL in a PIE
========================================
```
What packages need to be upgraded or what needs to be done to fix this issue?

Comment (ejgallego) [reply coq-core.opam:25]: No idea @JasonGross , I'm sorry, I wish I could be of more help but unfortunately I'm totally booked these days.

But that's a bit my point, very unfortunately, we don't have the resources in the Coq team these to be able to promise to help with there Coq-unrelated issues on a timely manner :( And on the other hand, our tooling / build systems needs a much needed refresh which we cannot delay more and will actually solve several dozen of issues with the current setup [such as rebases / bisects building properly, quick compilation, compilation caching, reduced maintenance, modularity, etc...]

I'd suggest tho you get in touch with Ralf Treinen and discuss.ocaml if you need Debian + OCaml help as I'm sure they will be much more helpful that we can [there is also OCaml discord and slack which is quite active]

Comment (ejgallego) [reply coq-core.opam:25]: The build failure is surprising tho as the Debian packages for 4.11.1 work well; likely you need to update binutils, but this is a good example of why this crosses Coq's boundaries, updating binutils does impact almost _all_ packages in a Debian system so that is way out of our league.

Personally I'd abandon Debian packages as a CI platform and switch to Docker / Nix.

Comment (JasonGross) [reply coq-core.opam:25]: Strangely enough, the build works fine in trusty, xenial, bionic, and focal (all of which are older than groovy) and hisrute (which is newer than groovy). Since groovy EOL's in July 2021, I can probably ignore this failure.

Comment (JasonGross) [reply coq-core.opam:25]: I've made a post on [ocaml.discuss](https://discuss.ocaml.org/t/building-ocaml-4-11-1-on-ubuntu-20-10/7370/1), though it got hidden by a spam filter.  Is this the best platform for this?

Comment (ejgallego) [reply coq-core.opam:25]: > I've made a post on [ocaml.discuss](https://discuss.ocaml.org/t/building-ocaml-4-11-1-on-ubuntu-20-10/7370/1), though it got hidden by a spam filter. Is this the best platform for this?

Ups ! I guess a moderator will take of it soon and approve your account, but indeed it is the best place to ask IMHO.

File: clib/cPath.mli
Comment (ejgallego) [reply clib/cPath.mli:18]: Fixed thanks!

File: clib/cPath.mli
Comment (ejgallego) [reply clib/cPath.mli:12]: Made it `private`, so it is a good start IMHO [and Path.t -> string is not so risky IMHO], it is a bit annoying to use, but hopefully when we consolidate the loadpath handling the current api will be much improved.

File: coq-stdlib.opam
Comment (ejgallego) [reply coq-stdlib.opam:11]: Updated, thanks.

File: coq-core.opam
Comment (ejgallego) [reply coq-core.opam:11]: Updated, let me know what you think.

File: test-suite/misc/printers.sh
Comment (ejgallego) [reply test-suite/misc/printers.sh:4]: Thanks, done.

File: coq-core.opam
Comment (ejgallego) [reply coq-core.opam:27]: Resolved, that was a stale thing as we had to disable opam file generation.

File: coq.opam
Comment (ejgallego) [reply coq.opam:27]: Umm, good point; only point here would be to have to install the `dune-package` file, will ask.

**Comment by ejgallego** [comment]
Thanks to all, I think only the Nix issue do remain, and a few other nits that IMHO shouldn't be blocking.

> The Nix error is probably due to the test-suite being run after installation. This use case might be broken by the current state of the PR.

I think we may be doing something wrong with the way Nix installs OCaml libs, like the meta or our `..` hack not working. The test suite is just calling `ocamlfind query coq-core` and this is not working for Nix for some reason.

@Zimmi48 @vbgl any help here would be extremely welcome.

File: coq-core.opam
Comment (Zimmi48) [inline coq-core.opam:36]: I thought this step was broken with Dune < 2.8...

File: coq-doc.opam
Comment (Zimmi48) [inline coq-doc.opam:25]: What justifies this change?

File: plugins/ltac/dune
Comment (Zimmi48) [inline plugins/ltac/dune:19]: Any reason to leave this commented code in here?

File: coq-core.opam
Comment (Zimmi48) [reply coq-core.opam]: LGTM.

**Comment by Zimmi48** [comment]
One thing that was concluded in the Coq Call and not yet implemented here IINM is to improve the error message of `coqtop` when it cannot find the stdlib to indicate to users that they probably didn't install the correct package.

File: lib/envars.ml
Comment (vbgl) [inline lib/envars.ml:208]: This is fishy: when the build is not local but the test-suite is run `coqlib ()` returns the path of the source. In this case, the sibling directory `coq-core` either does not exist or is unrelated.

File: coq-core.opam
Comment (ejgallego) [reply coq-core.opam:36]: Yes, sorry, that was a fluke on my part [I temporary enable opam file generation to generate the opam files, then I tweak]

**Comment by ejgallego** [comment]
> One thing that was concluded in the Coq Call and not yet implemented here IINM is to improve the error message of coqtop when it cannot find the stdlib to indicate to users that they probably didn't install the correct package.

Ups, I had written somewhere but I forgot, indeed will fix, thanks.

File: lib/envars.ml
Comment (ejgallego) [reply lib/envars.ml:208]: Umm, at least in the regular CI `coqlib` will return the installed path so indeed while hacky this should work, how do you call the test suite to get coqlib to point to the source in non-local build?

File: plugins/ltac/dune
Comment (ejgallego) [reply plugins/ltac/dune]: Resolved.

File: coq-doc.opam
Comment (ejgallego) [reply coq-doc.opam:25]: Was a mistake, fixed thanks.

File: coq.opam
Comment (ejgallego) [reply coq.opam:27]: I am gonna leave it like this, as this could be useful for example to have the coq package run the test suite, but indeed we may want to further tweak; but indeed the dune subst, etc... is required to have proper pinning of Opam packages, so I think that due to this bug we should bump to 2.8 which while not ideal, it is convenient IMHO.

File: lib/envars.ml
Comment (ejgallego) [reply lib/envars.ml:209]: Problem identified but was due to incorrect setting of COQLIB in coq makefile, actually the OSX job did also incorrectly set OCAMLPATH.

**Comment by ejgallego** [comment]
Ok, thanks a lot for the help folks, this is looking more and more ready, CI should be fully green if I didn't do any silly mistake.

I think I have covered all the points raised, there is still a bit of hackery in some parts regarding paths which were already hacky but IMO the best approach there is to merge this, merge #13617 , and once that is done do the cleanup, and for example, remove all of our OCaml path handling in favor of ocamlfind.

**Comment by ejgallego** [comment]
Note an important change since the last version [reflected in the changelog], we now require paths for `-prefix` to be absolute, after reviewing how paths are manipulated IMHO this is by far the best and safest solution, as opposed to actually checking lifting w.r.t. cd is done everywhere.

File: lib/envars.ml
Comment (Zimmi48) [inline lib/envars.ml:136]: This is not user-friendly enough. The message should contain an additional hint like "If you don't understand what this means, it probably means that you installed the wrong package (coq-core instead of coq)."

**Review by Zimmi48 — COMMENTED** [review]
Now it seems there's something off in the macOS build.

**Comment by ejgallego** [comment]
> Now it seems there's something off in the macOS build.

I did a silly mistake setting ocamlpath, should work now.

Going to fix the comment, will push in a second.

File: lib/envars.ml
Comment (ejgallego) [reply lib/envars.ml]: New version.

**Comment by Zimmi48** [comment]
There still is a failure in the test-suite on macOS.

**Comment by ejgallego** [comment]
> There still is a failure in the test-suite on macOS.

It turns out you need to perform `install-byte` too for the test-suite to be able to run from an installed Coq, Nix was already doing that. Crossing the fingers now.

**Comment by Zimmi48** [comment]
Will merge when CI passes.

**Comment by Zimmi48** [comment]
@coqbot merge now

**Review by Zimmi48 — APPROVED** [review]
@coqbot merge now

**Comment by ejgallego** [comment]
Thanks to all! Let me know of any issue / question on Zulip.

**Pull request merged**
